lib/Dialect/BGV/IR/BGVDialect.cpp

```diff
@@ -110,7 +113,19 @@ LogicalResult Relinearize::inferReturnTypes(
   auto x = cast<lwe::RLWECiphertextType>(adaptor.getInput().getType());
   inferredReturnTypes.push_back(lwe::RLWECiphertextType::get(
       ctx, x.getEncoding(),
-      lwe::RLWEParamsAttr::get(ctx, 2, x.getRlweParams().getRing())));
+      lwe::RLWEParamsAttr::get(ctx, 2, x.getRlweParams().getRing()),
+      x.getUnderlyingType()));
+  return success();
+}
+
+LogicalResult ExtractOp::verify() {
+  auto inputTy = cast<lwe::RLWECiphertextType>(getInput().getType());
+  auto tensorTy = dyn_cast<RankedTensorType>(inputTy.getUnderlyingType());
+  if (!tensorTy) {
+    return emitOpError() << "input RLWE ciphertext type must have a ranked "
+                            "tensor as its underlying_type, but found "
+                         << inputTy.getUnderlyingType();
+  }
```

Comment:
I think it also need to check that the output underlying type is the `tensorTy.getElementType()` (according to the comment in the TD file)

Reply:
Good catch!